### PR TITLE
[DESK] Multi-monitor support improvement

### PR DESCRIPTION
**Pull request by KyleKatarn31**
## Purpose

_ This PR will attempt to improve support of multi-monitor configurations in display properties .cpl as the current implementation (maked as "incomplete") assume a givent X-resolution (1280) for all screen, regardless of the real x-resolution.
Current code is a "dummy" code. The purpose of this proposal is to move one step further without inducing regressions.

## Proposed changes

_ Use of the "actual" resolution of all "previous" screen when enumerating all displays monitors in order to compute the correct offset.

Post change (non-regression)
image

### CODE REVIEW

**Comment by JoachimHenze** [comment]
1.) This code is malformatted regarding indentation.
2.) Please describe what case exactly this code aims to fix. Otherwise I oppose committing this.

**Comment by KyleKatarn31** [comment]
1 - Ident fixed
2 - Description improved.

File: dll/cpl/desk/settings.c [hunk 1]
```diff
@@ -378,10 +378,12 @@ SettingsOnInitDialog(IN HWND hwndDlg)
         pMonitors = (PMONSL_MONINFO)HeapAlloc(GetProcessHeap(), 0, sizeof(MONSL_MONINFO) * Result);
         if (pMonitors)
         {
-            DWORD hack = 1280;
+            DWORD offset = 0;
             for (i = 0; i < Result; i++)
             {
-                pMonitors[i].Position.x = hack * i;
+                if (i>1) offset = offset + pMonitors[i-1].Size.cx;
```
Comment (binarymaster) [inline dll/cpl/desk/settings.c:384]:
```suggestion
                if (i > 1)
                {
                    offset += pMonitors[i - 1].Size.cx;
                }
```

**Comment by KyleKatarn31** [comment]
Thanks for the code improvement. OK for me.
I have no multi-monitor VirtualBox to propose a before/after screenshot, sorry.

**Comment by learn-more** [comment]
Then how did you validate that this does that it is supposed to do?

**Comment by KyleKatarn31** [comment]
At least by analysis it's at least as good and likely better than the one making an assumption on resolution (which can indeed be too small or too large)

**Comment by KyleKatarn31** [comment]
Done by test now on a single screen configuration.

**Comment by katahiromz** [comment]
You shouldn't use "last" or "latest" for the title of content. Trailing space has to be deleted.

**Comment by KyleKatarn31** [comment]
> 
> 
> You shouldn't use "last" or "latest" for the title of content. Trailing space has to be deleted.

Sure :-)
Trailing space where ?

File: dll/cpl/desk/settings.c [hunk 2]
```diff
@@ -378,10 +378,15 @@ SettingsOnInitDialog(IN HWND hwndDlg)
...
         if (pMonitors)
         {
-            DWORD hack = 1280;
+            DWORD offset = 0;
             for (i = 0; i < Result; i++)
             {
-                pMonitors[i].Position.x = hack * i;
+                if (i > 1)
+                {
+                    offset += pMonitors[i - 1].Size.cx;
```
Comment (katahiromz) [inline dll/cpl/desk/settings.c:386]: Add:
```
offset += pMonitors[i].Size.cx;
```
After this line:
```
pMonitors[i].Flags = 0;
```

Comment (KyleKatarn31) [reply dll/cpl/desk/settings.c:386]: Very nice optimization. I take it !

File: dll/cpl/desk/settings.c [hunk 3]
```diff
@@ -378,14 +378,15 @@ SettingsOnInitDialog(IN HWND hwndDlg)
...
+            DWORD offset = 0;
             for (i = 0; i < Result; i++)
             {
-                pMonitors[i].Position.x = hack * i;
+                pMonitors[i].Position.x = offset;
                 pMonitors[i].Position.y = 0;
                 pMonitors[i].Size.cx = pData->DisplayDeviceList->CurrentSettings->dmPelsWidth;
                 pMonitors[i].Size.cy = pData->DisplayDeviceList->CurrentSettings->dmPelsHeight;
                 pMonitors[i].Flags = 0;
+				offset += pMonitors[i].Size.cx;
```
Comment (katahiromz) [inline dll/cpl/desk/settings.c:389]: Indentation.

Comment (KyleKatarn31) [reply dll/cpl/desk/settings.c]: Fixed...

**Comment by KyleKatarn31** [comment]
> 
> 
> I'm ok with the changes, but they should be tested in multi-monitor environment either in VM or on real hardware, with Windows XP or Windows Server 2003.

I understand but :
- Who has such a setup ? (not me)
- I seriously doubt that "pre-change" code has been tested and seriously validated in such configuration.

Do you see any particular risk in doing this enhancement ?

**Comment by giawa** [comment]
I ran both the current version of DESK and the proposed patch in a VM.  On the left is the current version, on the right is the patch.  It is an improvement in some cases, but quite a bit worse in others.  The multi-monitor support in general is pretty poor, and could use a general overhaul.  For example, clicking a monitor doesn't change the slider to the resolution of that window, etc.

**Improvement:**
improvement

**Regression:**
overlap

**Comment by KyleKatarn31** [comment]
> 
> 
> I ran both the current version of DESK and the proposed patch in a VM. On the left is the current version, on the right is the patch. It is an improvement in some cases, but quite a regression in others. The multi-monitor support in general is pretty poor, and could use a overhaul. For example, clicking a monitor doesn't change the slider for the resolution of that window, etc.
> 
> **Improvement:**
> improvement
> 
> **Regression:**
> overlap

Thank you. I will rework. 
May I have the details of the configuration which raises regressions ? (selected resolutions of both monitors)

**Comment by giawa** [comment]
> 
> 
> > I ran both the current version of DESK and the proposed patch in a VM. On the left is the current version, on the right is the patch. It is an improvement in some cases, but quite a regression in others. The multi-monitor support in general is pretty poor, and could use a overhaul. For example, clicking a monitor doesn't change the slider for the resolution of that window, etc.
> 
> Thank you. I will rework.
> May I have the details of the configuration which raises regressions ? (selected resolutions of both monitors)

Any situation where monitor 2 was a lower resolution that monitor 1.  In this example, monitor 1 was 1280x1024, and monitor 2 was 800x600.

**Comment by KyleKatarn31** [comment]
@binarymaster : Would you please set the "WIP" tag on this one ?

**Comment by Saibamen** [comment]
@KyleKatarn31: GitHub added some improvements into Draft Pull Requests. Cypress uses it, check example: https://github.com/cypress-io/cypress/pull/3130

Docs: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request

**Comment by KyleKatarn31** [comment]
@binarymaster : please add the WIP tag as the code will need some rework (not only more tests)

**Comment by binarymaster** [comment]
We are planning to get rid of WIP label and use draft PRs in this case. I converted your PR to draft.

**Comment by KyleKatarn31** [comment]
Very clear. Thanks

**Comment by KyleKatarn31** [comment]
Brings more regressions than benefits. No interest until ROS gets a functionnal multi-monitor support. Abandonned.

**Pull request closed without merging**